File: 프로그래머스/lv2/87946. 피로도/피로도.py

answer = 0

def dfs(k, dungeons, visit, cnt):
    global answer
    if cnt > answer:
        answer =cnt
        
    for idx, d in enumerate(dungeons):
        if not visit[idx] and k >= d[0]:
            visit[idx] = 1
            dfs(k-d[1], dungeons, visit, cnt+1)
            visit[idx] = 0

def solution(k, dungeons):
    global answer
    
    visit = [0 for _ in range(len(dungeons))]
    dfs(k, dungeons, visit, 0)
        
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.js


function solution(k, dungeons) {
    var answer = 0;
    const n = dungeons.length;
    const visit = Array.from({length: n}, ()=> 0);
    
    function dfs(k, cnt) {
        answer = Math.max(answer, cnt);
    
        dungeons.forEach((el, idx) => {
            if (visit[idx] === 0 && k >= el[0]){
                visit[idx] = 1;
                dfs(k-el[1], cnt + 1);
                visit[idx] = 0;
            }
        })
    }
    dfs(k, 0);
    
    return answer;
}